invenio_records_restapi/restful.py

```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Invenio.
+# Copyright (C) 2015 CERN.
+#
+# Invenio is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public License as
+# published by the Free Software Foundation; either version 2 of the
+# License, or (at your option) any later version.
+#
+# Invenio is distributed in the hope that it will be
+# useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Invenio; if not, write to the
+# Free Software Foundation, Inc., 59 Temple Place, Suite 330, Boston,
+# MA 02111-1307, USA.
+#
+# In applying this license, CERN does not
+# waive the privileges and immunities granted to it by virtue of its status
+# as an Intergovernmental Organization or submit itself to any jurisdiction.
+
+"""REST API resources."""
+
+from flask import jsonify, make_response, request, url_for
+
+from flask.ext.restful import Resource, abort
+
+from invenio_db import db
+
+from invenio_records.api import Record
+
+from jsonpatch import JsonPatchException, JsonPointerException
+
+from sqlalchemy.orm.exc import NoResultFound
+
+
+def record_self_link(record, **kwargs):
+    """Creates self link to a given record.
+
+    :Parameters:
+        - `record` (Record): record to which the generated link will point.
+        - `**kwargs`: additional parameters given to flask.url_for.
+    :Returns: link pointing to the given record.
+    :Returns Type: str
+    """
+    # '.' is needed to reference the current blueprint context
+    return url_for('.' + RecordAPI.api_endpoint,
+                   record_id=record.model.id, **kwargs)
+
+
+def request_wants_json():
+    """Test if the request accept type accepts 'application/json'.
+    :Returns: True if the request's accept type contains 'application/json'
+    else False.
+    :Returns Type: bool
+    """
+    best = request.accept_mimetypes \
+        .best_match(['application/json'])
+    return best == 'application/json'
+
+
+def record_to_response(record):
+    """Build a flask response from the given record, formatting it in an
+    accepted format.
+    :Returns: A flask response.
+    :Returns Type: Response
+    """
+    if not request_wants_json():
+        # requested mime format not supported
+        abort(406)
+
+    # FIXME: use a formatter instead once it is implemented
+    self = record_self_link(record, _external=True)
+    response = jsonify({
+        'id': record.model.id,
+        'data': record,
+        'links': {
+            'self': self
+        }
+    })
+    response.headers['location'] = self
+    return response
+
+
+class RecordsRootAPI(Resource):
```

Comment:
We should probably have a clear naming convention for collections and item resources:

```
RecordsRootAPI
RecordAPI
```

Previously in e.g. deposit I think I used e.g.:

```
DepositionListResource
DepositionItemResource
```

Not sure I really like either of them :-)